docs/advanced-solidus/state-machines.mdx

```diff
@@ -1,211 +1,88 @@
-that can produce a transition between valid states. They're local to the state machine component. On
-the other hand, bus events can be published and consumed anywhere within the system and, per se,
-have nothing to do with the state machines.
-
-:::
-
-As noted in the [event bus](../customization/subscribing-to-events.mdx) guide, you can leverage the
-event bus to hook into core events. That's helpful when you need to perform something in response to
-a change in the system, but your logic is orthogonal (i.e., decoupled) to the main flow. Transitions
-between state machine states are good candidates to become hotspots where tangential logic is
-triggered.
-
-For instance, you might want to update your ERP or send an SMS when a payment is marked as
-completed. The cleaner way to do that is to publish an event when that happens and then subscribe to
-it. First, you need to override the `#complete` event method on `Spree::Payment` (see
-the [overrides section](../customization/customizing-the-core.mdx#using-overrides) for the required
-setup code):
-
-```ruby title="app/overrides/my_store/publish_payment_completed.rb"
-# frozen_string_literal: true
-
-module MyStore
-  module PublishPaymentCompleted
-    def complete
-      super.tap do |result|
-        Spree::Bus.publish(:payment_completed, payment: self) if result
-      end
-    end
-
-    ::Spree::Payment.prepend self
-  end
-end
-```
-
-Then, after [registering the new event](../customization/subscribing-to-events.mdx#custom-events),
-you can [create a subscriber](../customization/subscribing-to-events.mdx#subscription-to-events) for
-it.
-
-:::caution
-
-Why not register a new `after_transition` hook instead of overriding the event method (
-i..e, `#complete` in the example above)?
-
-State machine transitions (including their `after_` hooks) are wrapped within a database
-transaction. By overriding the method instead of using `after_transition`, we publish the bus event
-only after the transaction has been committed to the DB.
-
-Event subscribers should **always** be decoupled from the main transaction, so there's no point in
-blocking database access until your subscribers have finished running. In fact, it's a bad practice:
-a failed subscriber will roll back the whole DB transaction and potentially leave your system in an
-inconsistent state.
-
-:::
-
-:::info
-
-Ideally, Solidus would publish events for every state machine transition out of the box. Our event
-bus is fairly new and we're still working on it, but we'll get there eventually! In the meantime,
-you can check`Spree::Bus.registered_events` for the complete list of events that are already
-published.
+It's better to be conservative when customizing state machines. Try to apply the smallest possible
+set of changes, and if possible, avoid changing the defined states. You're dealing with the core of
+the domain model: large changes could branch out in unanticipated ways!
 
```

Comment:
After reading this paragraph, it's still not clear when to use overrides vs replace an entire state machine. I think we should be more clear about it to avoid confusion.

Reply:
I think you're right. I rephrased it and added examples for each case. Please, take a look! 🙂

Reply:
I don't know if I read it wrong but I still can't find a proper explanation about when to use overrides vs replace an entire state machine. Maybe you were referring to "changing existing" vs "adding orthogonal behavior" instead?

Reply:
@kennyadsl, I want to clarify the difference between modifying the same domain vs. adding orthogonal behavior. Replacing the entire state machine is only for advanced usage, and it can encompass both kinds of modifications. Does it make sense?

Reply:
Yes, right. You can add orthogonal behavior in the replaced state machine. 👍 

Still, I think we should help people understand when it's time to replace the state machine entirely. Do we have something like that stated anywhere?

Reply:
I think it's worth keeping to be honest, and we should address those things in the more generic explanation about how to customize what's provided. I was thinking to something like:

> The customizations explained above allow implementing surgical-precision changes to the state machine flows but sometimes, for more deep behavioral modifications, you may need to change extensive parts of how they act. In that case, Solidus allows replacing the entire state machine with something custom. Be aware that this comes with drawbacks... etc etc.

Just wanted to make it clear upfront, otherwise, people will see the how-to but won't know when to use that method before reading all of its content. Also, although I think that explanation makes sense there, I'm not sure it is in line with what [Diátaxis suggest](https://diataxis.fr/how-to-guides/#don-t-explain-concepts): that's why I mentioned adding a link to the main section where this same thing would be explained.

Reply:
Thanks, @kennyadsl, and sorry for not getting what you meant. I think it's a great change. I also felt uneasy with that piece of explanation in the how-to guide. We're now linking to it, which is what Diátaxis recommends. That's what I did:

- I added something very similar to your paragraph to the main section.
- I moved the warning from the how-to guide just after the new paragraph in the main section.
- I added a link from the how-to guide to the main section in the first paragraph within an added `as it comes with its drawbacks` fragment.